tests/shadowcasting_test.cpp

```diff
@@ -407,199 +408,591 @@ static constexpr float O = LIGHT_TRANSPARENCY_SOLID;
 static constexpr float V = LIGHT_TRANSPARENCY_OPEN_AIR;
 static constexpr float X = LIGHT_TRANSPARENCY_SOLID;
 
-static const point ORIGIN( 65, 65 );
+static const tripoint ORIGIN( 65, 65, 11 );
 
 struct grid_overlay {
-    std::vector<std::vector<float>> data;
-    point offset;
+    std::vector<std::vector<std::vector<float>>> data;
+    std::vector<std::vector<std::vector<bool>>> floor;
+    tripoint offset;
     float default_value;
+    bool default_floor = true;
 
     // origin_offset is specified as the coordinates of the "camera" within the overlay.
     grid_overlay( const point &origin_offset, const float default_value ) {
         this->offset = ORIGIN - origin_offset;
         this->default_value = default_value;
     }
+    grid_overlay( const tripoint origin_offset, const float default_value ) {
+        this->offset = ORIGIN - origin_offset;
+        this->default_value = default_value;
+    }
 
-    int height() const {
+    int depth() const {
         return data.size();
     }
-    int width() const {
+    int height() const {
         if( data.empty() ) {
             return 0;
         }
         return data[0].size();
     }
+    int width() const {
+        if( data.empty() || data[0].empty() ) {
+            return 0;
+        }
+        return data[0][0].size();
+    }
+    tripoint get_max() const {
+        return offset + tripoint( width(), height(), depth() );
+    }
 
-    float get_global( const point &p ) const {
+    float get_global( const tripoint &p ) const {
         if( p.y >= offset.y && p.y < offset.y + height() &&
-            p.x >= offset.x && p.x < offset.x + width() ) {
-            return data[ p.y - offset.y ][ p.x - offset.x ];
+            p.x >= offset.x && p.x < offset.x + width() &&
+            p.z >= offset.z && p.z < offset.z + depth() ) {
+            return data[ p.z - offset.z ][ p.y - offset.y ][ p.x - offset.x ];
```

Comment:
Might be worth an `inclusive_cuboid` to simplify some of these bounds checks.
Also might be worth delegating to `get_local(p - offset)`?

Reply:
Correction: I meant `half_open_cuboid`.